### PR TITLE
Task6

### PR DESCRIPTION
**Pull request by paulius27**
Weather forecast functionality.

### CODE REVIEW

File: WeatherApp/BL/Validation/Validation.cs [hunk 1]
```diff
@@ -1,7 +1,22 @@
...
     public class Validation : IValidation
     {
+        private readonly IConfiguration _config;
+
+        private int _forecastMinDays, _forecastMaxDays;
+
+        public Validation(IConfiguration config)
+        {
+            _config = config;
+            _forecastMinDays = int.Parse(_config["WeatherForecast:MinDays"]);
```
Comment (MaximGomon) [inline WeatherApp/BL/Validation/Validation.cs]: Best approach is to use TryParse

File: WeatherApp/ConsoleApp/Commands/ICommand.cs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+﻿namespace ConsoleApp.Commands;
+
+public interface ICommand
+{
+    public Task Execute();
```
Comment (MaximGomon) [inline WeatherApp/ConsoleApp/Commands/ICommand.cs]: "public" is not needed on interface methods

File: WeatherApp/DAL/GeocodingRepository.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+using System.Net.Http;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace DAL
+{
+    public class GeocodingRepository : IGeocodingRepository
+    {
+        private readonly string _apiKey;
+        private readonly HttpClient _httpClient = new HttpClient();
```
Comment (MaximGomon) [inline WeatherApp/DAL/GeocodingRepository.cs]: HttpClient needs to be disposed after use. Please use "using" keyword and move HttpClient to the body of method GetCoordinatesByCityNameAsync

File: WeatherApp/DAL/GeocodingRepository.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            _apiKey = apiKey;
+        }
+
+        public async Task<Coordinates> GetCoordinatesByCityNameAsync(string cityName)
+        {
+            using var response = await _httpClient.GetAsync($"http://api.openweathermap.org/geo/1.0/direct?q={cityName}&limit=1&APPID={_apiKey}");
+            response.EnsureSuccessStatusCode();
+            var responseBody = await response.Content.ReadAsStringAsync();
+
+            var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
```
Comment (MaximGomon) [inline WeatherApp/DAL/GeocodingRepository.cs:27]: JsonSerialized is case insensitive by default

File: WeatherApp/DAL/GeocodingRepository.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+            var geocoding = JsonSerializer.Deserialize<List<GeocodingResponse>>(responseBody, options).First();
```
Comment (MaximGomon) [inline WeatherApp/DAL/GeocodingRepository.cs]: Please use FirstOrDefault and check for a "null"

File: WeatherApp/DAL/IGeocodingRepository.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+﻿using DAL.Models;
+using System.Threading.Tasks;
+
+namespace DAL
+{
+    public interface IGeocodingRepository
+    {
+        public Task<Coordinates> GetCoordinatesByCityNameAsync(string cityName);
```
Comment (MaximGomon) [inline WeatherApp/DAL/IGeocodingRepository.cs]: remove "public" access modifier

File: WeatherApp/DAL/WeatherRepository.cs [hunk 7]
```diff
@@ -26,5 +31,29 @@ public async Task<double> GetTemperatureByCityNameAsync(string cityName)
 
             return weather.Main.Temp;
         }
+
+        public async Task<IEnumerable<WeatherForecast>> GetForecastByCoordinatesAsync(Coordinates coordinates, DateTime startDate, DateTime endDate)
+        {
+            using var response = await _httpClient.GetAsync($"https://api.open-meteo.com/v1/forecast?latitude={coordinates.Latitude.ToString(CultureInfo.InvariantCulture)}&longitude={coordinates.Longitude.ToString(CultureInfo.InvariantCulture)}&daily=temperature_2m_max,temperature_2m_min&timezone=auto&start_date={startDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture)}&end_date={endDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture)}");
```
Comment (MaximGomon) [inline WeatherApp/DAL/WeatherRepository.cs]: The same issue with HttpClient. Please create client under "using" construction

File: WeatherApp/DAL/WeatherRepository.cs [hunk 8]
```diff
@@ -26,5 +31,29 @@ public async Task<double> GetTemperatureByCityNameAsync(string cityName)
...
+            response.EnsureSuccessStatusCode();
+            var responseBody = await response.Content.ReadAsStringAsync();
+
+            var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
```
Comment (MaximGomon) [inline WeatherApp/DAL/WeatherRepository.cs:41]: Not needed because this is default options

File: WeatherApp/BL/Validation/Validation.cs (hunk 1 above)
Comment (paulius27) [reply WeatherApp/BL/Validation/Validation.cs]: TryParse is now used.

File: WeatherApp/ConsoleApp/Commands/ICommand.cs (hunk 2 above)
Comment (paulius27) [reply WeatherApp/ConsoleApp/Commands/ICommand.cs]: Removed "public" access modifier from interface methods.

File: WeatherApp/DAL/GeocodingRepository.cs (hunk 3 above)
Comment (paulius27) [reply WeatherApp/DAL/GeocodingRepository.cs]: HttpClient is now created using IHttpClientFactory and disposed via "using".

File: WeatherApp/DAL/GeocodingRepository.cs (hunk 5 above)
Comment (paulius27) [reply WeatherApp/DAL/GeocodingRepository.cs]: FirstOrDefault is now used.

File: WeatherApp/DAL/IGeocodingRepository.cs (hunk 6 above)
Comment (paulius27) [reply WeatherApp/DAL/IGeocodingRepository.cs]: Removed "public" access modifier from interface methods.

File: WeatherApp/DAL/WeatherRepository.cs (hunk 7 above)
Comment (paulius27) [reply WeatherApp/DAL/WeatherRepository.cs]: HttpClient is now created using IHttpClientFactory and disposed via "using".

**Review by MaximGomon — APPROVED** [review]

**Pull request merged**